src/crimson/os/seastore/btree/fixed_kv_node.h

```diff
@@ -712,125 +130,76 @@ struct FixedKVInternalNode
   }
   // Must be identical with FixedKVInternalNode(ptr) after on_fully_loaded()
   explicit FixedKVInternalNode(extent_len_t length)
-    : FixedKVNode<NODE_KEY>(CAPACITY, length) {}
+    : FixedKVNode<NODE_KEY>(length),
+      ParentNode<node_type_t, NODE_KEY>(CAPACITY) {}
   FixedKVInternalNode(const FixedKVInternalNode &rhs)
-    : FixedKVNode<NODE_KEY>(rhs) {
+    : FixedKVNode<NODE_KEY>(rhs),
+      ParentNode<node_type_t, NODE_KEY>(rhs) {
     this->set_layout_buf(this->get_bptr().c_str());
   }
 
-  bool have_children() const final {
-    return true;
-  }
-
-  bool is_leaf_and_has_children() const final {
-    return false;
-  }
-
-  uint16_t get_node_split_pivot() final {
+  uint16_t get_node_split_pivot() {
     return this->get_split_pivot().get_offset();
   }
 
   void prepare_commit() final {
-    if (this->is_initial_pending()) {
-      if (this->is_rewrite()) {
-	this->set_children_from_prior_instance();
-      }
-      this->copy_children_from_stable_sources(
-	[this](base_t &node, uint16_t pos) {
-	  ceph_assert(node.get_type() == this->get_type());
-	  auto &n = static_cast<this_type_t&>(node);
-	  return n.iter_idx(pos);
-	}
-      );
-      if (this->is_rewrite()) {
-	this->reset_prior_instance();
-      } else {
-	this->adjust_ptracker_for_children();
-      }
-      assert(this->validate_stable_children());
-      this->copy_sources.clear();
-    }
-  }
-
-  bool is_child_stable(
-    op_context_t<NODE_KEY>,
-    uint16_t pos,
-    NODE_KEY key) const final {
-    ceph_abort("impossible");
-    return false;
-  }
-  bool is_child_data_stable(
-    op_context_t<NODE_KEY>,
-    uint16_t pos,
-    NODE_KEY key) const final {
-    ceph_abort("impossible");
-    return false;
-  }
-
-  bool validate_stable_children() final {
-    LOG_PREFIX(FixedKVInternalNode::validate_stable_children);
-    if (this->children.empty()) {
-      return false;
-    }
-
-    for (auto i : *this) {
-      auto child = (FixedKVNode<NODE_KEY>*)this->children[i.get_offset()];
-      if (child && child->range.begin != i.get_key()) {
-	SUBERROR(seastore_fixedkv_tree,
-	  "stable child not valid: child {}, child meta{}, key {}",
-	  *child,
-	  child->get_node_meta(),
-	  i.get_key());
-	ceph_abort();
-	return false;
-      }
-    }
-    return true;
+    parent_node_t::prepare_commit();
   }
 
   virtual ~FixedKVInternalNode() {
     if (this->is_valid() && !this->is_pending()) {
-      if (this->range.is_root()) {
-	ceph_assert(this->root_block);
-	unlink_phy_tree_root_node<NODE_KEY>(this->root_block);
-      } else {
-	ceph_assert(this->is_parent_valid());
-	auto parent = this->template get_parent_node<FixedKVNode<NODE_KEY>>();
-	auto off = parent->lower_bound_offset(this->get_meta().begin);
-	assert(parent->get_key_from_idx(off) == this->get_meta().begin);
-	assert(parent->children[off] == this);
-	parent->children[off] = nullptr;
-      }
+      this->RootNode<RootBlock, node_type_t>::destroy();
+      this->child_node_t::destroy();
```

Comment:
Following another comment, if `ChildNode` and `RootNode` can be properly abstracted, there is no need to call both methods from different classes.

And it might be confusing that here is calling both methods, and some other places uses a different pattern:
https://github.com/ceph/ceph/blob/4ab2c4653e1bdfb44d3ebca4b4fd2670bf0246e4/src/crimson/os/seastore/btree/fixed_kv_node.h#L578-L582

Reply:
> And it might be confusing that here is calling both methods, and some other places uses a different pattern:
> 
> https://github.com/ceph/ceph/blob/4ab2c4653e1bdfb44d3ebca4b4fd2670bf0246e4/src/crimson/os/seastore/btree/fixed_kv_node.h#L578-L582

How about we follow this pattern for all cases and turn all `if (me.is_root())` in linked tree nodes into `assert(me.is_root())`?